### PR TITLE
refactor: move uriSegments to app directory

### PR DESCRIPTION
**Pull request by alexgoff**
This is phase 2 of site migration from pages to app directory: everything that isn't the homepage.

[https://ag-516.dev.rubinepo.dev/](https://ag-516.dev.rubinepo.dev/)

## Testing
This change is all encompassing so the following attempts to test to major types of content on the site.

- [Generic page](https://ag-516.dev.rubinepo.dev/about)
- [Event](https://ag-516.dev.rubinepo.dev/events/summer-stars-teaching-astronomy-with-rubin-observatory)
- [Glossary term](https://ag-516.dev.rubinepo.dev/for-educators/glossary/hubble-plot)
- [News post](https://ag-516.dev.rubinepo.dev/news/rubins-lsst-camera-is-complete)
- [Press release](https://ag-516.dev.rubinepo.dev/news/camera-arrives-chile)
- [Search results](https://ag-516.dev.rubinepo.dev/search?search=mirror)
- [Slideshow](https://ag-516.dev.rubinepo.dev/slideshows/making-a-telescope-mirror)
- [Staff profile](https://ag-516.dev.rubinepo.dev/explore/staff/ardis-herrold)
- [Dynamic content](https://ag-516.dev.rubinepo.dev/news)

### Visual
- Check side-by-side with production instance. Look for layout shifts, font changes, color changes, etc.
- Check for hydration errors (will pop up in dev mode)
- Check for any obviously wrong behavior

### Metadata

Each page should have

- A title in the format of [page] | Rubin Observatory
- An optional description, falls back to the sitewide description
- An opengraph/twitter image if there is a hero image on the page

### Preview
Should preview with draft content in Craft and in web-page

### Revalidation
Should revalidate, including disabling the page to receive a 404, and reenabling clearing the 404 page (test with devtools closed so browser cache is not disabled)

### Localization
Should show content in current locale
Should direct user to preferred locale
Switcher should switch between English and Spanish (hopefully faster than previously)
lang attribute on HTML root should change with locale

### Error page
Should go to error page if a page component throws an error. This is trickier to test, but React's dev-tools can force components into an errored state image

### 404 page
Should go to 404 page if an invalid URI is entered (just type "Dfasdflasdkfja;sldfkjsdlf"). It should also attempt to show the page in the user's preferred locale
On other pages:

### CODE REVIEW

File: app/[locale]/[...uriSegments]/page.tsx [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    return {
+      locale,
+      uriSegments: uri.split("/"),
+    };
+  });
+}
+
+const sectionMap = {
+  events: EventPageTemplate,
+  // galleryItems: GalleryPageTemplate,
```
Comment (alexgoff) [inline app/[locale]/[...uriSegments]/page.tsx:115]: for now I am excluding the gallery and user profile pages since both are unused.

File: app/[locale]/[...uriSegments]/page.tsx [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+  const entrySectionType = await getEntrySectionByUri(
+    uri,
+    locale,
+    previewToken
+  );
+
+  // Handle 404 if there is no data
+  if (!entrySectionType) {
+    notFound();
```
Comment (alexgoff) [inline app/[locale]/[...uriSegments]/page.tsx:145]: the previous static file purging has been removed. App directory uses cached data instead of static built pages so there is no pre-built content to purge in a 404 scenario

File: app/[locale]/layout.tsx [hunk 3]
```diff
@@ -49,8 +49,6 @@ export const generateStaticParams = () => {
   });
 };
 
-export const dynamicParams = false;
```
Comment (alexgoff) [inline app/[locale]/layout.tsx:52]: `dynamicParams = false` restricts routing to only the statically generated params. After the homepage migration this was `en` and `es` which worked fine. Now we have a mix of statically generated + dynamically rendered pages. It is not documented, but setting `dynamicParams = false` sets this for all child routes too which would cause pages like the news posts to 404.

File: app/api/app-revalidate/route.ts [hunk 4]
```diff
@@ -24,7 +27,16 @@ export async function GET(request: NextRequest): Promise<NextResponse> {
...
+    languages.forEach((locale) => {
+      const parts: Array<string> = uri === CRAFT_HOMEPAGE_URI ? [] : [uri];
+      if (locale !== fallbackLng) {
+        parts.unshift(locale);
+      }
+
+      revalidatePath(`/${parts.join("/")}`);
+    });
+
+    revalidateTag(tags.globals);
```
Comment (alexgoff) [inline app/api/app-revalidate/route.ts:39]: revalidate globals on any revalidate so stuff like navigation changing is picked up by all pages

File: components/global/Body/index.js
Comment (alexgoff) [inline components/global/Body/index.js:1]: deprecation, yay!

File: components/templates/SlideshowPage/index.tsx [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+const SlideshowPage: FunctionComponent<{
+  section: string;
+  data: PageEntry;
+  locale: string;
+  previewToken?: string;
+}> = async ({ data: { uri }, section, locale, previewToken }) => {
+  const data = await getSlideshowDataByUri(uri, locale, previewToken);
+
+  return <SlideshowPageClient {...{ data, section }} />;
```
Comment (alexgoff) [inline components/templates/SlideshowPage/index.tsx:13]: slideshows had their own custom data fetch, this has been moved into a server file for the template

File: lib/api/canto/index.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
+import { fallbackLng } from "@/lib/i18n/settings";
+import { resizeCantoImage, ValidCantoSizes } from "./resize";
+
+const responsiveCantoSrc = (
```
Comment (alexgoff) [inline lib/api/canto/index.ts:4]: move over some Canto helpers from the investigations that we can start using for pages whose opengraph image is a Canto asset

File: lib/api/client/query.ts [hunk 7]
```diff
@@ -18,12 +20,21 @@ if (
 const queryAPI = async <Query, Variables extends AnyVariables = AnyVariables>({
   query,
   variables,
+  fetchOptions: inputFetchOptions,
```
Comment (alexgoff) [inline lib/api/client/query.ts:23]: expose fetch options to through our API client, mainly for tagging fetches for revalidation

File: lib/api/fragments/educator-page.js [hunk 8]
```diff
@@ -38,11 +38,6 @@ fragment educatorPageFragmentFull on pages_educatorPages_Entry {
         ${getImageFields("crop", 900, 550)}
       }
     }
-    openGraphImage: hero {
```
Comment (alexgoff) [inline lib/api/fragments/educator-page.js:41]: opengraph images are fetched from a single metadata method instead of as part of the entry's query

File: lib/styles/font.ts [hunk 9]
```diff
@@ -6,6 +6,7 @@ const SourceSansPro = Source_Sans_3({
   style: ["normal", "italic"],
   subsets: ["latin"],
   display: "swap",
+  fallback: ["system-ui"],
```
Comment (alexgoff) [inline lib/styles/font.ts:9]: Lighthouse was mad that there was no fallback for our font

File: pages/api/revalidate.js [hunk 10]
```diff
@@ -33,17 +30,16 @@ async function handler(req, res) {
...
-      await fetch(`${BASE_URL}/api/app-revalidate?${searchParams.toString()}`);
-      return res.status(200).json({ revalidated: true });
-    } else {
-      await res.revalidate(`/${query.uri}`);
-      return res.status(200).json({ revalidated: true });
-    }
-  } catch (error) {
-    console.error(error);
+  const searchParams = new URLSearchParams(query);
+  const response = await fetch(
```
Comment (alexgoff) [inline pages/api/revalidate.js:34]: forward all revalidates to the app directory route handlers. Next step will be to delete this endpoint and rename the app directory one to take over

**Comment by ericdrosas87** [comment]
Some feedback:

View/preview query param is `null` (when I click "View" on the entry page):
http://localhost:3000/news/coating-primary-mirror?preview=null

For translation:
When in Spanish the title shows “null” for the second part (where Vera Rubin Observatory should be, like "Historia | null")

Errors:
Forcing error on router causes white error page
Forcing error on link component causes Next.js error in dev mode
Basically I was never able to get to the error page
